### ISSUE TITLE
npm not calculating correct "wanted" version

### ISSUE THREAD

**Issue by jd-carroll** [npm/npm#20210]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [x] npm is producing an incorrect install.
  - [x] npm is doing something I don't understand.
  - [x] npm is producing incorrect or undesirable behavior.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?
`npm` is not producing the correct `wanted` version when running `npm outdated`.  The result is that any call to `npm update` will not produce the desired install.

#### How can the CLI team reproduce the problem?
In an empty directory run the following steps:

- `$ npm init -y`
- `$ npm install axios@0.15.3`
- Validate that your `package.json` is consistent with the following:

```json
{
  "name": "your-dir",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "axios": "^0.15.3"
  }
}
```

- `$ npm outdated` should produce the following table:

```
Package  Current  Wanted  Latest  Location
axios     0.15.3  0.15.3  0.18.0  your-dir
```

**faceplant** - As you should be able to see in the table, the wanted version is listed as `0.15.3` but the latest version is `0.18.0`.  **This is problematic because the pacakge.json lists the version as inclusive of minor & patch release updates and is listed as: `^0.15.3`**.  

If the version in the package.json was listed as `~0.15.3`, this would be perfectly acceptable behavior.

### supporting information:

 - `npm -v` prints: 5.6.0
 - `node -v` prints: v8.9.4
- Windows 10

**Comment by kevinrambaud** [comment]
Do you have a package-lock file ? If yes, did you checked which version was installed for your `axios` dependency in the lock file ?

**Comment by jd-carroll** [comment]
By default it will generate a `package-lock.json` since this is npm 5.  The package-lock looks like this:

```
{
  "name": "your-dir",
  "version": "1.0.0",
  "lockfileVersion": 1,
  "requires": true,
  "dependencies": {
    "axios": {
      "version": "0.15.3",
      "resolved": "https://artifacts.internal.com/artifactory/api/npm/npm-repo/axios/-/axios-0.15.3.tgz",
      "integrity": "sha1-LJ1jiy4ZGgjqHWzJiOrda6W9wFM=",
      "requires": {
        "follow-redirects": "1.0.0"
      }
    },
    "debug": {
      "version": "2.6.9",
      "resolved": "https://artifacts.internal.com/artifactory/api/npm/npm-repo/debug/-/debug-2.6.9.tgz",
      "integrity": "sha1-XRKFFd8TT/Mn6QpMk/Tgd6U2NB8=",
      "requires": {
        "ms": "2.0.0"
      }
    },
    "follow-redirects": {
      "version": "1.0.0",
      "resolved": "https://artifacts.internal.com/artifactory/api/npm/npm-repo/follow-redirects/-/follow-redirects-1.0.0.tgz",
      "integrity": "sha1-jjQpjL0uF28lTv/sdaHHjMhJ/Tc=",
      "requires": {
        "debug": "2.6.9"
      }
    },
    "ms": {
      "version": "2.0.0",
      "resolved": "https://artifacts.internal.com/artifactory/api/npm/npm-repo/ms/-/ms-2.0.0.tgz",
      "integrity": "sha1-VgiurfwAvmwpAd9fmGF4jeDVl8g="
    }
  }
}
```

Note: I created this example from our office (hence the different `resolved`).  I am able to reproduce the same behavior at home on my Windows 10 laptop.

**Comment by kevinrambaud** [comment]
I already noticed that in the past. If I recall correctly, it might be due to the fact that packages that didn't reach the 1.0.0 release do not behave like the one that are considered as stable.

For example, try to install a lodash version such as `npm i lodash@3.9.2`, run the `npm outdated` command, check the result and `npm update`

**Comment by jiripospisil** [comment]
> As you should be able to see in the table, the wanted version is listed as 0.15.3 but the latest version is 0.18.0. This is problematic because the pacakge.json lists the version as inclusive of minor & patch release updates and is listed as: ^0.15.3.

As @kevinrambaud pointed out, this is expected - pre 1.0.0 versions behave differently.

```js
const semver = require("semver")

> semver.satisfies("0.18.0", "^0.15.3")
false

> semver.satisfies("1.18.0", "^1.15.3")
true
```

The reasoning behind it is that pre 1.0.0 versions are not considered stable and thus any version change is potentially breaking. You have at least two options:

1. Install every pre 1.0.0 version manually
2. Modify the range to include everything up to 1.0.0 non inclusive
  ```js
  > semver.satisfies("0.18.0", ">=0.15.3 <1.0.0")
  true
  ```

For future reference, you can see the matching versions using [npm semver calculator](https://semver.npmjs.com/).

**Comment by kevinrambaud** [comment]
I think it would be a good idea to explicitly mention the pre-1.0.0 behavior in the `npm update` documentation, this is a source of confusion for a lot of people.

**Comment by jd-carroll** [comment]
Agreed, I've been working with node+npm for multiple years and this is the first I am running into it... I'll work on putting together a PR to fix documentation for `npm update` + `npm outdated`.

**Comment by mcclowes** [comment]
@jd-carroll can this issue be closed now? Did you put together that PR?
